pkg/reconciler/ingress/ingress.go

```diff
@@ -578,3 +583,19 @@ func getLBStatus(gatewayServiceURL string) []v1alpha1.LoadBalancerIngressStatus
 func enablesAutoTLS(ctx context.Context) bool {
 	return config.FromContext(ctx).Network.AutoTLS
 }
+func isIstioInjectionEnabledForNs(r *BaseIngressReconciler, ra ReconcilerAccessor, ia v1alpha1.IngressAccessor) (bool, error) {
```

Comment:
We've been putting off detecting if we're running in a mesh or not. Is this check really exhaustive? I think you'd also need to check whether the sending/receiving service has istio injection enabled/disabled specifically.

Moreover, you'd need to check if the injection-webhook is even installed and if so if it is enabled via ConfigMap.

Long story short: Is detection really necessary or is there a way to work around it? What's the harm in adding the entry in a non-mesh case?

Reply:
Current behavior is that short and full FQDN entries are always added to spec.host regardless of whether Istio sidecar injection is enabled or not for non-mesh virtual services.
```
spec:
  gateways:
  - cluster-local-gateway
  hosts:
  - helloworld-test-image.default
  - helloworld-test-image.default.svc
  - helloworld-test-image.default.svc.cluster.local
```
For mesh services, only full FQDN is added.
```
spec:
  gateways:
  - mesh
  hosts:
  - helloworld-test-image.default.svc.cluster.local
```
If sidecar injection is enabled, mesh virtual service is used for route traffic. And the service couldn't be found when short FQDN is used.

My thought is to add short FQDN entries to mesh service only if service visibility is "cluster-local" and Istio-injection is enabled for namespace.

Reply:
sidecar.istio.io/inject: "false" could only take effect when "Istio-injection" label for namespace is enabled.

Reply:
I'm unclear why we need to check for istio injection at all? What's the harm of always adding it?

Reply:
There should be no harm in always adding, since VirtualService exposed to `mesh` is only applied to the sidecar proxies anyway.

However, I believe Istio already does the short name expansions you mentioned, when VirtualService is for `mesh` (which was the reason why we split them in the first place).  

And looking at https://testgrid.knative.dev/serving#istio-1.2-mesh&exclude-non-failed-tests=20, I see that the tests TestServiceToServiceCall/short TestServiceToServiceCall/shortest are passing.  

Same with https://testgrid.knative.dev/serving#istio-1.1-mesh&exclude-non-failed-tests=20 .  

In Istio 1.0 (which is end of life now), there are some bugs with the way ExternalName are handled.  That has been fixed in 1.1 and 1.2.

Reply:
@tcnghia @markusthoemmes 
Tested with istio 1.1.7 and tested again, it works without this PR changes.
By default, outboundTrafficPolicy of Istio is ALLOW_ANY.
Then switch outboundTrafficPolicy to REGISTRY_ONLY and run the test again.
This time it didn't work and reported 404 again.
For knative serving, could we will use default value of istio.outboundTrafficPolicy?
Or do we need to use REGISTRY_ONLY?
https://istio.io/docs/tasks/traffic-management/egress/egress-control/#envoy-passthrough-to-external-services

Reply:
Is the failed test in the partial mesh case (either client and server are in a mesh, but not both)?  If `REGISTRY_ONLY` means to only include things in the mesh, I think we only need to test partial mesh connectivity for ALLOW_ANY.

If the test also failed in the case of both client and server are in a mesh and REGISTRY_ONLY is enabled we should definitely open an issue to fix that.  Btw, do the changes here fix the issue when REGISTRY_ONLY is used?  

Thanks

Reply:
Both client and server are in a mesh and server is with label "serving.knative.dev/visibility: cluster-local". Then tests failed if Istio-injection is enabled for the namespace. 
If label "Istio-injection" for namespace is not enabled, the tests could also pass.
This changes could fix the problem.